specs/transport/car/carv2/index.md

```diff
@@ -0,0 +1,177 @@
+---
+title: "CARv2 Specification"
+eleventyNavigation:
+  synopsys: "Specification for the CARv2 format"
+---
+
+# Specification: Content Addressable aRchives (CAR / .car) v2
+
+**Status: Draft**
+
+* [Summary](#summary)
+* [Format Description](#format-description)
+  * [Pragma](#pragma)
+  * [Header](#header)
+  * [Characteristics](#characteristics)
+  * [CARv1 data payload](#carv1-data-payload)
+  * [Index payload](#index-payload)
+  * [Index format](#index-format)
+    * [Format `0x300000`: IndexSorted](#format-0x300000-indexsorted)
+* [Implementations](#implementations)
+* [Test Fixtures](#test-fixtures)
+
+## Summary
+
+CARv2 is a minimal upgrade to the [CARv1](../carv1/index.md) format with the primary aim of adding an optional index within the format for fast random-access to blocks.
+
+CARv2 makes use of CARv1 by wrapping a properly formed CARv1 with a prefix containing a pragma and header, and a suffix containing the optional index data. Once the offset and length of the CARv1 bytes are determined using CARv2 parsing rules, an existing CARv1 decoder can be used to read the roots and `CID:Bytes` pairs. Likewise, a CARv1 encoder can be used to encode this data for wrapping by a CARv2 encoder.
+
+## Format Description
+
+CARv2 consists of:
+
+1. An 11-byte pragma (or sequence of "magic bytes") that identify the data as a CARv2 format.
+2. A header describing data payload offset and length and index offset within the container as well as a bitfield for setting various capabilities supported by the container.
+3. A standard CARv1 data payload, including standard CARv1 header and roots and sequence of `CID:Bytes` pairs.
+4. An optional index payload, which may be one of a number of supported index formats, allowing for fast lookups of blocks within the data payload in the form of an offset from the start of the container.
+
+The CARv2 format can be illustrated as follows:
+
+```
+| 11-byte fixed pragma | 40-byte header | optional padding | CARv1 data payload | optional padding | optional index payload
+```
+
+### Pragma
+
+The CARv2 version pragma (or "magic bytes") was chosen for compatibility with existing CARv1 parsers. CARv1 leads with a [DAG-CBOR](../../../codecs/dag-cbor/index.md) block, prefixed with a varint where the block contains the version and roots array:
```

Comment:
Being picky: would it make sense to eliminate all references to "magic", in favour of pure unadulterated logic? Would it add anything to introduce "magic bytes" as an alternative name for the pragma?

Reply:
There's no such thing as pure unadulterated logic!

But also, I was thinking of linking to https://en.wikipedia.org/wiki/List_of_file_signatures (which also calls them "Magic Bytes") to help make the point that this isn't a huge number of bytes for this purpose compared to other formats.

I've removed the first reference in the description above but left this one and added the link. Let me know if it still smells to you, I'm happy to remove as it's not a big deal.

Reply:
This is perfectly fine; please ignore me 😊